src/npctalk_rules.cpp

```diff
@@ -0,0 +1,309 @@
+#include "npctalk_rules.h"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <cstddef>
+#include <iterator>
+#include <list>
+#include <map>
+#include <memory>
+#include <ostream>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "cata_imgui.h"
+#include "dialogue.h"
+#include "npctalk.h"
+#include "ui_manager.h"
+#include "imgui/imgui.h"
+
+static std::map<cbm_recharge_rule, std::string> recharge_map = {
+    {cbm_recharge_rule::CBM_RECHARGE_ALL, "<ally_rule_cbm_recharge_all_text>" },
+    {cbm_recharge_rule::CBM_RECHARGE_MOST, "<ally_rule_cbm_recharge_most_text>" },
+    {cbm_recharge_rule::CBM_RECHARGE_SOME, "<ally_rule_cbm_recharge_some_text>" },
+    {cbm_recharge_rule::CBM_RECHARGE_LITTLE, "<ally_rule_cbm_recharge_little_text>" },
+    {cbm_recharge_rule::CBM_RECHARGE_NONE, "<ally_rule_cbm_recharge_none_text>" },
+};
+
+static std::map<cbm_reserve_rule, std::string> reserve_map = {
+    {cbm_reserve_rule::CBM_RESERVE_ALL, "<ally_rule_cbm_reserve_all_text>" },
+    {cbm_reserve_rule::CBM_RESERVE_MOST, "<ally_rule_cbm_reserve_most_text>" },
+    {cbm_reserve_rule::CBM_RESERVE_SOME, "<ally_rule_cbm_reserve_some_text>" },
+    {cbm_reserve_rule::CBM_RESERVE_LITTLE, "<ally_rule_cbm_reserve_little_text>" },
+    {cbm_reserve_rule::CBM_RESERVE_NONE, "<ally_rule_cbm_reserve_none_text>" },
+};
+
+static std::map<combat_engagement, std::string> engagement_rules = {
+    {combat_engagement::NONE, "<ally_rule_engagement_none>" },
+    {combat_engagement::CLOSE, "<ally_rule_engagement_close>" },
+    {combat_engagement::WEAK, "<ally_rule_engagement_weak>" },
+    {combat_engagement::HIT, "<ally_rule_engagement_hit>" },
+    {combat_engagement::ALL, "<ally_rule_engagement_all>" },
+    {combat_engagement::FREE_FIRE, "<ally_rule_engagement_free_fire>" },
+    {combat_engagement::NO_MOVE, "<ally_rule_engagement_no_move>" },
+};
+
+static std::map<aim_rule, std::string> aim_rule_map = {
+    {aim_rule::WHEN_CONVENIENT, "<ally_rule_aim_when_convenient>" },
+    {aim_rule::SPRAY, "<ally_rule_aim_spray>" },
+    {aim_rule::PRECISE, "<ally_rule_aim_precise>" },
+    {aim_rule::STRICTLY_PRECISE, "<ally_rule_aim_strictly_precise>" },
+};
```

Comment:
Unless these need to be internally sorted or the key isn't hashable, it's generally preferred to use `unordered_map`

Reply:
Right, I misunderstood you and tried to apply this. The ordering is in fact important, as it's used for the hotkey wrapping. std::map::upper_bound lets us press the hotkey and move through the map